Add Nintendo Switch as tier 3 target

jam1garner opened the pull request:
[Relevant Zulip Discussion](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Upstreaming.20Nintendo.20Switch.20Support/near/253445503)

This is the first step towards working on incrementally adding support for the Nintendo Switch. After this lands @leo60228 and I will work on ensuring further work is cleared from a legal perspective before continuing on to work on an allocator and porting libstd.

The plan is to keep these changes small and incremental enough so as to not cause unneeded burden on reviewers by submitting a single large patch, as was felt to be the case last attempt at upstreaming (#74567).

All this specific patch does is add the target itself without and std support, which has been tested on-device and is working as expected.

Designated Target Maintainers:
* @leo60228 
* @jam1garner

jam1garner committed:
Add Nintendo Switch tier 3 target

leo60228 committed:
Don't use host linker for switch

leo60228 committed:
Remove unnecessary linker args

leo60228 committed:
Add linker script for switch

leo60228 committed:
Don't build std for switch

rust-highfive commented:
r? @wesleywiser

(rust-highfive has picked a reviewer for you, use r? to override)

rust-highfive commented:
:warning: **Warning** :warning:

* These commits modify **compiler targets**. (See the [Target Tier Policy](https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html).)

nagisa commented:
Please copy the [list of target requirements](https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html#tier-3-target-policy) and verify all of the requirements are satisfied.

An example from one of the previous PRs: https://github.com/rust-lang/rust/pull/88529#issuecomment-909515760

petrochenkov commented on compiler/rustc_target/src/spec/aarch64_nintendo_switch.rs:
```diff
@@ -0,0 +1,36 @@
+
+    opts.pre_link_args.insert(LinkerFlavor::Lld(LldFlavor::Ld), vec![]);
+
+    opts.post_link_args.insert(LinkerFlavor::Lld(LldFlavor::Ld), vec![]);
```
A number of fields here use values that are already used by default (see `impl Default for TargetOptions`), they can all be removed.

jam1garner committed:
Remove unneeded options from Nintendo Switch target

leo60228 commented:
> - A tier 3 target must have a designated developer or developers (the "target maintainers") on record to be CCed when issues arise regarding the target. (The mechanism to track and CC such developers may evolve over time.)

Yep, me and @jam1garner.

> - Targets must use naming consistent with any existing targets; for instance, a target for the same CPU or OS as an existing Rust target should use the same name for that CPU or OS. Targets should normally use the same names and naming conventions as used elsewhere in the broader ecosystem beyond Rust (such as in other toolchains), unless they have a very good reason to diverge. Changing the name of a target can be highly disruptive, especially once the target reaches a higher tier, so getting the name right is important even for a tier 3 target.
>   - Target names should not introduce undue confusion or ambiguity unless absolutely necessary to maintain ecosystem compatibility. For example, if the name of the target makes people extremely likely to form incorrect beliefs about what it targets, the name should be changed or augmented to disambiguate it.

We spent a good amount of time bikeshedding this, but `aarch64-nintendo-switch` is consistent with `armv6k-nintendo-3ds` and `mipsel-sony-psp`. One consideration was targets built on top of C libraries, but there's some precedent here in the form of `aarch64-apple-ios` and `aarch64-apple-ios-sim`.

> - Tier 3 targets may have unusual requirements to build or use, but must not create legal issues or impose onerous legal terms for the Rust project or for Rust developers or users.
>   - The target must not introduce license incompatibilities.
>   - Anything added to the Rust repository must be under the standard Rust license (`MIT OR Apache-2.0`).
>   - The target must not cause the Rust tools or libraries built for any other host (even when supporting cross-compilation to the target) to depend on any new dependency less permissive than the Rust licensing policy. This applies whether the dependency is a Rust crate that would require adding new license exceptions (as specified by the `tidy` tool in the rust-lang/rust repository), or whether the dependency is a native library or binary. In other words, the introduction of the target must not cause a user installing or running a version of Rust or the Rust tools to be subject to any new license requirements.
>   - If the target supports building host tools (such as `rustc` or `cargo`), those host tools must not depend on proprietary (non-FOSS) libraries, other than ordinary runtime libraries supplied by the platform and commonly used by other binaries built for the target. For instance, `rustc` built for the target may depend on a common proprietary C runtime library or console output library, but must not depend on a proprietary code generation library or code optimization library. Rust's license permits such combinations, but the Rust project has no interest in maintaining such combinations within the scope of Rust itself, even at tier 3.
>   - Targets should not require proprietary (non-FOSS) components to link a functional binary or library.
>   - "onerous" here is an intentionally subjective term. At a minimum, "onerous" legal/licensing terms include but are *not* limited to: non-disclosure requirements, non-compete requirements, contributor license agreements (CLAs) or equivalent, "non-commercial"/"research-only"/etc terms, requirements conditional on the employer or employment of any particular Rust developers, revocable terms, any requirements that create liability for the Rust project or its developers or users, or any requirements that adversely affect the livelihood or prospects of the Rust project or its developers or users.

The code in this PR definitely shouldn't pose any issue. Before making PRs for libstd, we plan to consult the core team and foundation (as discussed on Zulip), but none of the explicitly mentioned issues apply.

> - Neither this policy nor any decisions made regarding targets shall create any binding agreement or estoppel by any party. If any member of an approving Rust team serves as one of the maintainers of a target, or has any legal or employment requirement (explicit or implicit) that might affect their decisions regarding a target, they must recuse themselves from any approval decisions regarding the target's tier status, though they may otherwise participate in discussions.
>   - This requirement does not prevent part or all of this policy from being cited in an explicit contract or work agreement (e.g. to implement or maintain support for a target). This requirement exists to ensure that a developer or team responsible for reviewing and approving a target does not face any legal threats or obligations that would prevent them from freely exercising their judgment in such approval, even if such judgment involves subjective matters or goes beyond the letter of these requirements.

:thumbsup:

> - Tier 3 targets should attempt to implement as much of the standard libraries as possible and appropriate (`core` for most targets, `alloc` for targets that can support dynamic memory allocation, `std` for targets with an operating system or equivalent layer of system-provided functionality), but may leave some code unimplemented (either unavailable or stubbed out as appropriate), whether because the target makes it impossible to implement or challenging to implement. The authors of pull requests are not obligated to avoid calling any portions of the standard library on the basis of a tier 3 target not implementing those portions.

See discussion on Zulip. We're planning on adding support incrementally to make review easier for both ourselves and reviewers.

> - The target must provide documentation for the Rust community explaining how to build for the target, using cross-compilation if possible. If the target supports running tests (even if they do not pass), the documentation must explain how to run tests for the target, using emulation if possible or dedicated hardware if necessary.

Currently, there's not much specific to the target. https://github.com/aarch64-switch-rs/cargo-nx works with `--triple=aarch64-nintendo-switch`.

> - Tier 3 targets must not impose burden on the authors of pull requests, or other developers in the community, to maintain the target. In particular, do not post comments (automated or manual) on a PR that derail or suggest a block on the PR based on a tier 3 target. Do not send automated messages or notifications (via any medium, including via `@`) to a PR author or others involved with a PR regarding a tier 3 target, unless they have opted into such messages.
>   - Backlinks such as those generated by the issue/PR tracker when linking to an issue or PR are not considered a violation of this policy, within reason. However, such messages (even on a separate repository) must not generate notifications to anyone involved with a PR who has not requested such notifications.

:thumbsup:

> - Patches adding or updating tier 3 targets must not break any existing tier 2 or tier 1 target, and must not knowingly break another tier 3 target without approval of either the compiler team or the maintainers of the other tier 3 target.
>   - In particular, this may come up when working on closely related targets, such as variations of the same architecture with different features. Avoid introducing unconditional uses of features that another variation of the target may not have; use conditional compilation or runtime detection, as appropriate, to let each target run code supported by that target.

Assuming CI passes, I don't see any reason to think this would be an issue.

leo60228 commented:
Also, for some relevant information that was posted on Zulip but wasn't copied here, this PR just adds a target definition. Future PRs for libstd will be pure Rust: we haven't signed any NDAs for the official SDK, and we don't plan to rely on unofficial C/C++ SDKs.

bjorn3 commented on compiler/rustc_target/src/spec/aarch64_nintendo_switch_linker_script.ld:
```diff
@@ -0,0 +1,83 @@
+  .mod0 : {
+    KEEP(crt0.nso.o(.data.mod0))
+    KEEP(crt0.nro.o(.data.mod0))
+    KEEP(crt0.lib.nro.o(.data.mod0))
```
What are those object files? Where can you get them from?

leo60228 replied:
Actually, I'm not sure these are necessary. cc @XorTroll (who I believe wrote the original linker script?)

leo60228 replied:
Yeah, this was just a mistake. I don't think these have existed for ages, I've removed them.

jam1garner replied:
These were object files generated from assembly in libtransistor (permissive OSS for building homebrew on the switch) for starting up the executable. They aren't needed under Rust's build system. As leo said he's removed them from the linker script, although that wasn't doing anything before.

leo60228 committed:
Remove obsolete crt0 references in linker script

sgeos commented:
If this is an NDA compliant port, please drop me a line on the English language official forums under "Off-Topic/Target Triple for Rust std" or "Off-Topic/Rust".  If this is a homebrew initiative, an indication as such would be appreciated.

leo60228 commented:
> Also, for some relevant information that was posted on Zulip but wasn't copied here, this PR just adds a target definition. Future PRs for libstd will be pure Rust: we haven't signed any NDAs for the official SDK, and we don't plan to rely on unofficial C/C++ SDKs.

@sgeos Does this answer your question? I agree that this probably should've been in the PR description, but I didn't think of that while we were writing it and I can't edit it.

sgeos commented:
> > Also, for some relevant information that was posted on Zulip but wasn't copied here, this PR just adds a target definition. Future PRs for libstd will be pure Rust: we haven't signed any NDAs for the official SDK, and we don't plan to rely on unofficial C/C++ SDKs.
> 
> @sgeos Does this answer your question? I agree that this probably should've been in the PR description, but I didn't think of that while we were writing it and I can't edit it.

The short answer is yes.

Aaron1011 commented on src/bootstrap/config.rs:
```diff
@@ -308,7 +308,7 @@ pub struct Target {
     pub fn from_triple(triple: &str) -> Self {
         let mut target: Self = Default::default();
-        if triple.contains("-none") || triple.contains("nvptx") {
+        if triple.contains("-none") || triple.contains("nvptx") || triple.contains("switch") {
```
Off topic - could this be turned into a `no_std` flag in the target config file?

leo60228 replied:
The linker has the same issue with a `FIXME` for pulling the info from the target config file. I'm guessing that `bootstrap` doesn't have a way to access the actual target definition.

leo60228 replied:
In this PR, though, it's just temporary until std *is* implemented on Switch.

leo60228 commented:
For what it's worth, as a note on the naming, the Switch's operating system is a microkernel called Horizon. Horizon was *also* used on the 3DS, but it evolved quite a bit between the two consoles. `aarch64-nintendo-horizon` and `aarch64-unknown-horizon` were also considered, but with the acceptance of a 3DS target under the name `armv6k-nintendo-3ds` we decided to go with `aarch64-nintendo-switch` just for consistency.
One possible wrinkle here is that there are three projects reimplementing the Horizon microkernel:
* [mesosphere](https://github.com/Atmosphere-NX/Atmosphere/tree/master/mesosphere): complete C++ reimplementation targeting bug-for-bug compatibility, formally only supports the Switch but with a suitable bootloader could likely be ported to other Tegra X1 hardware with minimal changes to the kernel itself
* [SunriseOS](https://github.com/sunriseos/SunriseOS): Rust reimplementation targeting i386, no longer developed (IIRC the authors were making it as part of an operating systems course, which has concluded)
* [Saturnus](https://github.com/vbe0201/Saturnus): early Rust reimplementation, I believe it currently targets the Switch but is planned to be a general-purpose RTOS

These projects might want to use the same target, in which case the `aarch64-nintendo-switch` name *could* be misleading. I'm not sure this is an actual issue, though, and I think consistency with the 3DS takes priority.

jam1garner commented:
I think one relevant point about this being primarily designed for homebrew (I can't imagine an officially sanctioned game using this; if only because they need to dynamically link against the nnsdk), however we would like to keep the possibility for there to be another target designated for use when linking against the nnsdk. (an `aarch64-nintendo-switch-nnsdk`, if you will)

This is further complicated by the fact our plan for this target is to be a pure Rust toolchain (just interfacing with the syscall interface), which makes it a bit weird trying to specify a toolchain in the target. I think `aarch64-nintendo-switch` doesn't properly articulate it being for homebrew, but there isn't really a toolchain to append to it (like `-musl` on Linux targets). From a technical standpoint it makes sense as a 'default' as it's the least common denominator and would work in any context, but from a social standpoint much less so.

At risk of bikeshedding: If anyone has any thoughts on how this should be handled (even if just 'leave it alone'), I'd appreciate outside input

jam1garner commented:
> For what it's worth, as a note on the naming, the Switch's operating system is a microkernel called Horizon. Horizon was _also_ used on the 3DS, but it evolved quite a bit between the two consoles. `aarch64-nintendo-horizon` and `aarch64-unknown-horizon` were also considered, but with the acceptance of a 3DS target under the name `armv6k-nintendo-3ds` we decided to go with `aarch64-nintendo-switch` just for consistency.
> One possible wrinkle here is that there are three projects reimplementing the Horizon microkernel:
> 
>     * [mesosphere](https://github.com/Atmosphere-NX/Atmosphere/tree/master/mesosphere): complete C++ reimplementation targeting bug-for-bug compatibility, formally only supports the Switch but with a suitable bootloader could likely be ported to other Tegra X1 hardware with minimal changes to the kernel itself
> 
>     * [SunriseOS](https://github.com/sunriseos/SunriseOS): Rust reimplementation targeting i386, no longer developed (IIRC the authors were making it as part of an operating systems course, which has concluded)
> 
>     * [Saturnus](https://github.com/vbe0201/Saturnus): early Rust reimplementation, I believe it currently targets the Switch but is planned to be a general-purpose RTOS
> 
> 
> These projects might want to use the same target, in which case the `aarch64-nintendo-switch` name _could_ be misleading. I'm not sure this is an actual issue, though, and I think consistency with the 3DS takes priority.

I think it's also worth noting that mesosphere is the only one out of the three currently 'production ready' (so to speak) and can for all intents and purposes be considered only for use on the switch, so I think actual room for confusion from 'switch' in the name to be nonexistant.

sgeos commented:
> I think one relevant point about this being primarily designed for homebrew (I can't imagine an officially sanctioned game using this; if only because they need to dynamically link against the nnsdk), however we would like to keep the possibility for there to be another target designated for use when linking against the nnsdk. (an `aarch64-nintendo-switch-nnsdk`, if you will)
> 
> This is further complicated by the fact our plan for this target is to be a pure Rust toolchain (just interfacing with the syscall interface), which makes it a bit weird trying to specify a toolchain in the target. I think `aarch64-nintendo-switch` doesn't properly articulate it being for homebrew, but there isn't really a toolchain to append to it (like `-musl` on Linux targets). From a technical standpoint it makes sense as a 'default' as it's the least common denominator and would work in any context, but from a social standpoint much less so.
> 
> At risk of bikeshedding: If anyone has any thoughts on how this should be handled (even if just 'leave it alone'), I'd appreciate outside input

aarch64-nintendo-switch does not conflict with any target triples that have been discussed behind the NDA wall, but it still might be confusing.  aarch64-nintendo-switch-homebrew is entirely unambiguous.  Personally, I would love to see a target that can be used for commercial.  I am hoping this effort can be used as a base since you will probably move faster than me.  Not that any such target, if created, will remain behind the NDA wall, probably forever.

jam1garner commented:
>  Personally, I would love to see a target that can be used for licensed commercial development. I am hoping this effort can be used as a basis for an official or semi-official target since you will probably move faster than me. Note that any such target, if created, will likely remain behind the NDA wall, probably forever.

One thing I hope (after further legal discussion, because I believe this will be an even more dicey endeavor than the currently proposed homebrew target) is that once further development is done on this target there could be a non-NDA'd target that would purely be based on the exposed symbols of the Nintendo SDK.

The reason I believe this *could* be possible is that (please note before I continue that I am not under a Nintendo NDA, have never looked upon or been given information covered by a Nintendo NDA, or otherwise interacted with NDA materials) the Nintendo SDK exports quite a large portion of libc required to produce a working Rust libstd.

I have [experimented with this in the past for the purpose of modding games in the past](https://github.com/skyline-rs/rust) (consider the fact that when modding games you are subject to a programming environment equivalent to that of the game, so a lot of this work could very well be helpful to a toolchain targetting 'official' game development).

I would believe that at minimum the subset of libc exported (sockets, pthreads, and a limited file API) would be enough to produce a semi-useful target without \*any\* copyrighted materials being involved (as, while I am not a lawyer and this is ofc not legal advice, I do not believe Nintendo has the rights to copyright a subset of libc :P )

As to whether or not such a target could dynamically link against other symbols in the nnsdk within libstd is less clear to me, however, but at minimum (a) skyline-rs/rust is a quite functional starting place for those under NDA making games if they would like a starting point and (b) I am willing to answer any questions about it people have (this is not the place to do so but it's not hard to find me elsewhere).

While working with the Rust Foundation and/or Rust core team when working on the homebrew libstd implementation I intend to also inquire about this. Hopefully there is a course of action I can take to enable development in the open but ultimately not putting the Rust project in legal jeopardy is my #1 priority for either of the proposed targets.

jam1garner commented:
Also I should add that the current gameplan is that leo and I are currently assessing libstd implementation strategy and I plan on contacting both the foundation and the core team once both:
1. an implementation strategy is clear enough to present to legal and
2. this PR has merged

sgeos commented:
> > Personally, I would love to see a target that can be used for licensed commercial development. I am hoping this effort can be used as a basis for an official or semi-official target since you will probably move faster than me. Note that any such target, if created, will likely remain behind the NDA wall, probably forever.
> 
> One thing I hope @sgeos (after further legal discussion, because I believe this will be an even more dicey endeavor than the currently proposed homebrew target) is that once further development is done on this target there could be a non-NDA'd target that would purely be based on the exposed symbols of the Nintendo SDK.
> 
> The reason I believe this _could_ be possible is that (please note before I continue that I am not under a Nintendo NDA, have never looked upon or been given information covered by a Nintendo NDA, or otherwise interacted with NDA materials) the Nintendo SDK exports quite a large portion of libc required to produce a working Rust libstd.
> 
> I have [experimented with this in the past for the purpose of modding games in the past](https://github.com/skyline-rs/rust) (consider the fact that when modding games you are subject to a programming environment equivalent to that of the game, so a lot of this work could very well be helpful to a toolchain targetting 'official' game development).
> 
> I would believe that at minimum the subset of libc exported (sockets, pthreads, and a limited file API) would be enough to produce a semi-useful target without *any* copyrighted materials being involved (as, while I am not a lawyer and this is ofc not legal advice, I do not believe Nintendo has the rights to copyright a subset of libc :P )
> 
> As to whether or not such a target could dynamically link against other symbols in the nnsdk within libstd is less clear to me, however, but at minimum (a) skyline-rs/rust is a quite functional starting place for those under NDA making games if they would like a starting point and (b) I am willing to answer any questions about it people have (this is not the place to do so but it's not hard to find me elsewhere).
> 
> While working with the Rust Foundation and/or Rust core team when working on the homebrew libstd implementation I intend to also inquire about this. Hopefully there is a course of action I can take to enable development in the open but ultimately not putting the Rust project in legal jeopardy is my #1 priority for either of the proposed targets.

First of all, I am not an attorney.  If you have legal concerns, it is a good idea to consult an attorney.  Having said that, my understanding is that reverse trade secrets is legal.  Reverse engineering cannot be used as a defense for patent infringement, but there is precedent for interoperability and overturning silly patents.

The upshot is that reverse engineering whatever Nintendo is doing an implementing a target based on that knowledge is almost certainly legal, but there are a few things to be aware of.  First of all, after signing an NDA and being approved as a developer, access a lot of trade secret secrets is granted, technical and nontechnical.  People who have signed such an NDA are obligated not to share these trade secrets even if they stop developing for the system.  Anyone behind the NDA wall is legally unable to share trade secrets with you, but you are free to use publicly available information.  I am pretty sure that leaks are public information that can be used, but I would double check this.

If someone who has signed an NDA feed you information, that puts both you and that person at risk legally.  I have seen projects roll back functionality that is based on propriety information.  I think that this is a good idea if you get information that cannot be used legally.  To the extent you have not signed an NDA, you are free to work and share information openly, but you may not make of trade secrets provided by someone who has signed an NDA.  People who have signed an NDA are free to use reverse engineered work, license permitting, but they cannot share trade secrets.

As far as concrete detail go, I do not see any problem building a target that relies on reverse engineer symbols that are present on the device.  If you write a target that uses whatever subset of libc is exported, your aarch64 implementation may actually work on other aarch64 targets that export all of libc.  Furthermore, it sounds like it might also work on the 3DS for the people who are still interested in developing homebrew for it.  I think that working with the available libc symbols is the ideal strategy to use.

skade commented:
@Sgeos Thanks for the writeup and your view on this. From a project management perspective though, whether something is _legal or not_ is not the only concern. The problem is an intersection of the structure of the `libc` crate and legal interpretations. Because `libc` is one library spelling out all the targets, any action taken against one target - independent of the outcome - will impact all the others. That's the reason why we need to be careful and very solid in footing there. _Every target_ needs to rely on our reading to be correct and when the reading includes external documents and agreements, it gets problematic.

That would be very different if a `libc-switch` crate existed and could be linked instead of the `libc` crate.

sgeos commented:
@skade As far as external agreements go, my recommendation is to only allow people to contribute if A) they have no external agreements (NDA not signed, NDA trade secrets never seen, etc.), or B) they have explicit permission to contribute despite having signed an NDA.  My understanding is that external documentation based on reverse engineering should be OK, but I honestly have no idea what the Switch homebrew community is using.  Some of the other homebrew communities use extensive data sheets of technical information.

Does it make sense to create a `libc-switch` crate and conditionally link to it instead of the standard `libc` crate?  If so, then the crate not existing is a solvable problem.  Furthermore, implementing it may be an excellent strategy for approaching this kind of target.  Worst case scenario, this `libc-switch` crate can simply be removed if unforeseeable problems emerge.

nagisa commented:
I'm not sure we are in a position to verify that the contributors haven't signed some sort of an NDA, and drawing a line in the code for where such a verification should be made and where its not necessary seems infeasible at best.

skade commented:
@sgeos I'll write a full reply later, but just to make clear: the `libc-switch` crate was just a strawman, not a real idea and proposal. I just wanted to use it to illustrate how a separate scope may ease things.

leo60228 commented:
> My understanding is that external documentation based on reverse engineering should be OK, but I honestly have no idea what the Switch homebrew community is using. Some of the other homebrew communities use extensive data sheets of technical information.

Leaked materials definitely aren't used, though the public Tegra X1 TRM is frequently referenced for low-level details.

There's likely enough documentation for clean-room implementations, but if you consider having looked at any disassembly/decompilation to taint someone, you'll have a hard time finding developers. It's definitely a far better situation than the GameCube, where I've heard that chunks of the toolchain were directly taken from decompiler output...

For what it's worth, I think the fact that Nintendo hasn't taken action against Atmosphere (which has bug-compatible reimplementations of many system components) is good evidence that they wouldn't take action against a Rust port.

wesleywiser reviewed (changes requested):
Can you also add documentation for this target to `src/doc/rustc/src/platform-support` in the style of the existing [iOS simulator docs](https://github.com/rust-lang/rust/blob/master/src/doc/rustc/src/platform-support/aarch64-apple-ios-sim.md)?

leo60228 commented:
@wesleywiser I was actually planning on doing that but I mistakenly thought those were in a separate repo, I'll write some.

leo60228 committed:
Add docs for Switch target

leo60228 commented:
For what it's worth, I'm open to a target name *like* `aarch64-nintendo-switch-homebrew`, but that feels *way* too long and I don't have any better ideas... It's also not especially descriptive: it could also be ambiguous with a theoretical libnx target.

wesleywiser commented:
> Also I should add that the current gameplan is that leo and I are currently assessing libstd implementation strategy and I plan on contacting both the foundation and the core team once both:
> 
> 1. an implementation strategy is clear enough to present to legal and
> 2. this PR has merged

I think it would be helpful to go ahead and start this conversation now. From a technical perspective, the code here is fine and could be signed off on as it is. However, while it seems ok to me to merge this, I recognize that I am not a lawyer and don't really know what I'm talking about. 🙂 

I would prefer if we had some kind of agreed upon strategy here for the target with those folks. If, for example, we could not find an implementation path that allows for in-tree `libc`/`libcore`/`libstd` support for this target, that would place this target below Tier 3 support with no way to reach that level which seems undesirable long-term. In such a case, is having the target definition in-tree even useful? Hopefully, that is not the conclusion and we can provide robust support for the target but it would be good to clarify the situation at this time.

leo60228 commented:
For what it's worth: `libc` support isn't necessary, as this target wouldn't have a libc. `core` already works.

wesleywiser commented:
@leo60228 Ah yes, you're correct. I was just reading back over the comments and noticed that. I've updated my comment. Thanks for the correction!

> For what it's worth, I'm open to a target name like aarch64-nintendo-switch-homebrew, but that feels way too long and I don't have any better ideas... It's also not especially descriptive: it could also be ambiguous with a theoretical libnx target.

Perhaps something like `aarch64-nintendo-switch-raw` or `aarch64-nintendo-switch-freestanding` would be appropriate?

leo60228 commented:
I like `aarch64-nintendo-switch-freestanding`, but it's even better than `-homebrew`. I feel like `-raw` might not be especially clear, but I'm curious what other people's thoughts are there.

sgeos commented:
@leo60228 I guess I can throw this out there.  The "nx" codename for the Switch is considered public knowledge and it has been approved by Nintendo for use in open source projects.  Most people probably use it for things like conditional compilation.  On that note, there is officially sanctioned precedent for open source projects to target the Switch.

Having said that, if nx is used in the target triple, then it very much should be marked as unofficial to avoid confusion.  Even though nx has been cleared for open source use, official developers probably make up the bulk of people who are aware of the designation.  Everyone else probably expects Switch.  Something like `aarch64-nintendo-nx-homebrew` is a little shorter than spelling out Switch.  For serious work, the target triple can be captured in a file, so it should not be a problem if it is a little long.

leo60228 commented:
@sgeos That might make since, but it's pretty clearly inconsistent with the recently added 3DS target being `aarch64-nintendo-3ds` and not `aarch64-nintendo-ctr`.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3633168936) [(plain)](https://github.com/rust-lang/rust/commit/117f45c34c222a1c1f97db7a0114cf148beae287/checks/3633168936/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Set({"src/doc/edition-guide"}) not skipped for "bootstrap::doc::EditionGuide" -- not in ["src/tools/tidy"]
Rustbook (x86_64-unknown-linux-gnu) - edition-guide
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.16s
rustc/print.html:6688: broken link - `rustc/platform-support/aarch64-nintendo-switch.html`
rustc/platform-support.html:314: broken link - `rustc/platform-support/aarch64-nintendo-switch.html`
number of HTML files scanned: 28332
number of HTML redirects found: 8809
number of links checked: 1773526
number of links ignored due to external: 66202
```

</details>

leo60228 committed:
Add aarch64-nintendo-switch.md to SUMMARY.md

I can't think of any other reason CI might be failing, and I should've
done this anyway.

leo60228 commented:
As a point of comparison, `aarch64-nintendo-switch-freestanding` would be tied with `mipsisa64r6el-unknown-linux-gnuabi64` and `thumbv7neon-unknown-linux-musleabihf` as the longest target at 36 characters.

sgeos commented:
> > but people behind the NDA wall are free to use reverse engineered work, license permitting.
> 
> Precisely why I think this would actually be a good route to go. Having maintained an out-of-tree libstd, I believe that an attempt to use a target out-of-tree behind the NDA wall would be a worse experience, _even for the people who signed an NDA_. Updating downstream after `library/std/src/sys/` refactors is not fun.

@jam1garner For what it is worth, I really want to see a Switch target that can be kept up to date with rustup.  Despite my candid feedback, I really want to see this target come to fruition.

leo60228 commented:
After further thought and some bikeshedding with other people involved with
Rust on Switch homebrew efforts, I'm currently strongly leaning towards
aarch64-nintendo-switch-freestanding. If nobody has any objections to that,
I'll update the name soon.

On Fri, Sep 17, 2021, 12:50 PM Brendan A R Sechter ***@***.***>
wrote:

> but people behind the NDA wall are free to use reverse engineered work,
> license permitting.
>
> Precisely why I think this would actually be a good route to go. Having
> maintained an out-of-tree libstd, I believe that an attempt to use a target
> out-of-tree behind the NDA wall would be a worse experience, *even for
> the people who signed an NDA*. Updating downstream after
> library/std/src/sys/ refactors is not fun.
>
> @jam1garner <https://github.com/jam1garner> For what it is worth, I
> really want to see a Switch target that can be kept up to date with rustup.
> Despite my candid feedback, I really want to see this target come to
> fruition.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/88991#issuecomment-921940382>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB7X32OGCPLNBJ5KRFBIYJLUCNWXHANCNFSM5EDLWRAA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jam1garner commented:
> @jam1garner For what it is worth, I really want to see a Switch target that can be kept up to date with rustup. Despite my candid feedback, I really want to see this target come to fruition.

heh, glad we're in the same boat. I think it will be hard and will require a lot of work to ensure it doesn't pose any risk to the Rust project, but I think the benefits of stuff like "bevy developers can see if their audio library compiles for switch without jumping through 20 hoops" would be huge for Rust gamedev. I'd love to see a future where console developers can really feel like first class citizens here, even if it's gonna be tough to get there \:)

> I'm currently strongly leaning towards aarch64-nintendo-switch-freestanding. If nobody has any objections to that, I'll update the name soon.

I am also on board with this change, given @wesleywiser suggested the name I assume they are also good with.

sgeos commented:
> @jam1garner  heh, glad we're in the same boat. I think it will be hard and will require a lot of work to ensure it doesn't pose any risk to the Rust project, but I think the benefits of stuff like "bevy developers can see if their audio library compiles for switch without jumping through 20 hoops" would be huge for Rust gamedev. I'd love to see a future where console developers can really feel like first class citizens here, even if it's gonna be tough to get there :)

One of the biggest problems I have had is that it is difficult to talk to people in the Rust community and find help from behind the NDA wall.  I can talk about the hypothetical heavily NDA protected "smart toaster" I am developing for.  The problem is that standard operating system assumptions are broken at a high level (as mentioned above by @leo60228 only some of the libc symbols are exported), and I cannot get into details.

I really want to use standard Rust crates for commercial game development.  Keeping development in the open just makes things easier, but to a large extent only people who have *not* signed an NDA can do that.  It would be great to see published Switch games written in Rust!

leo60228 committed:
Rename aarch64-nintendo-switch to aarch64-nintendo-switch-freestanding

nellshamrell commented:
Hi @jam1garner and everyone else! Nell from the Rust Foundation here. I'm seeing a few different legal related questions throughout this thread. Could someone write up a summary of your questions for me? I'm happy to pass it on to the Foundation's attorney.

jam1garner commented:
> Hi @jam1garner and everyone else! Nell from the Rust Foundation here. I'm seeing a few different legal related questions throughout this thread. Could someone write up a summary of your questions for me? I'm happy to pass it on to the Foundation's attorney.

First off: Thank you! Saves me an email :)

As for a summary of initial questions:
1. `aarch64-nintendo-switch-freestanding` is a target based purely on reverse-engineered information pertaining to the system call interface of the Nintendo Switch, and contains no code or other copyrighted materials of Nintendo. It will be purely a from-scratch implementation of the Rust standard library using information obtained from reverse engineering. (One important resource being [Switchbrew](https://switchbrew.org/wiki/SVC), which does not accept contributions from individuals under NDA and does not allow copyrighted materials). Is this of an acceptable level confidence from a legal perspective to be upstreamed?
2. `aarch64-nintendo-switch-nnsdk` is another potential target that would be based around linking against the Nintendo SDK. The information needed to do this has not been obtained by any channel involving an NDA and has instead been achieved by observing the dynamic linking interface exposed by the 'nnsdk' binary present in the contents of every Nintendo game. Same question as last time: Is this of an acceptable level confidence from a legal perspective to be upstreamed? Two implementation options below:

    a. Linking against any nnsdk symbols dynamically, using no nintendo code and merely using information obtained via reverse engineering in order to understand how the interface can be used.
    b. If the above is too loose to upstream such a target and that the exposed interface is not something we can legally use: a majority of the needed interface exposed from the nnsdk is in fact just a subset of libc, and thus isn't specific to the switch nor Nintendo. If the only option for an nnsdk-based target is to merely target only that subset of libc, then a less functional libstd implementation could be made using only these libc functions.

With all of these, if the answer is only yes under certain conditions, knowing what those stipulations would be would be helpful.

Additional context:

Nobody working on any of the aforementioned targets has ever been under a Nintendo NDA, nobody working on this has looked at illegally obtained copyrighted materials of Nintendo, and nobody working on this has had information only available under an NDA disseminated to them.

If any more private contact is needed @nellshamrell feel free to pass along my email: jam@jam1.re, and leo can be cc'd if desired at leo@60228.dev

nellshamrell commented:
Thank you, @jam1garner! I've reached out to our counsel and will reach out for more context from you if they have more questions. I will report back as soon as I can!

bors commented:
:umbrella: The latest upstream changes (presumably #86191) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jam1garner commented:
@wesleywiser before I bothering fixing merge conflicts, are we waiting on legal comment before merging this PR, or are we considering this change far enough removed from anything remotely questionable that I can fix up the pull request so we can work on getting it merged?

wesleywiser commented:
I would prefer if we could have that comment from legal before proceeding. 

@nellshamrell Do you have an idea when we might hear back from the legal counsel?

nellshamrell commented:
@wesleywiser - we are still working on it, I hope to have an answer (or at least direction) for you in the next week or two

skade commented:
@jam1garner just wanted to leave a ping that I was in touch with @wesleywiser on this issue and we're working on a proper solution and a policy for this situation. We'll need some time to write it down and circle around, but would then get you involved to see it's workable and to your benefit.

Milo123459 commented:
Hey! Just had a quick question, say we compile the code to the Nintendo switch target, how do we actually run it?

jam1garner commented:
> @jam1garner just wanted to leave a ping that I was in touch with @wesleywiser on this issue and we're working on a proper solution and a policy for this situation. We'll need some time to write it down and circle around, but would then get you involved to see it's workable and to your benefit.

Sounds great.

> Hey! Just had a quick question, say we compile the code to the Nintendo switch target, how do we actually run it?

While not super relevant to the PR: this is left up to the user. The most conventional way is to aquire a development kit, but we also intend for this implementation (e.g. the freestanding target and the subject of this PR) to not be dependent on the nnsdk or be authored involving anyone under NDA—and thus would be able to run under a homebrew environment. However either target should be runnable under emulators in some form. 

One such example of code that would be compiled dynamically linking against the nnsdk would be mods for games—which would be able to run on a system running custom firmware without any copyrighted materials being involved in Rust/compilation itself, similar to how dll injection on Windows doesn't require the dll to include copyrighted materials.

For the freestanding implementation most testing will be performed under a homebrew environment while most nnsdk-based development (if permitted) will be tested by the switch equivalent of dll injection (for the curious: the keywords of interest are "layeredfs" and "Atmosphère custom firmware")

Hope that answers your question.

sgeos commented:
> Hey! Just had a quick question, say we compile the code to the Nintendo switch target, how do we actually run it?

Official devkits can be used to test on hardware during development.  If you are a licensed developer, feel free to reply to one of the Rust threads on the official forums.  I will be happy to collaborate via official channels.

If you are not an official developer, then you are probably looking at homebrew toolchains.  Presumably, the homebrew community has documented how they test on hardware.  Emulators are another option for testing homebrew software.

jam1garner commented:
@nellshamrell and @skade just checking in, nothing needed from our end currently? No rush just want to check on the status on things. Also if you need more real-time communication than email/github, feel free to shoot me a message on the rust-lang zulip as well

AlexApps99 commented:
Are there any updates on the blockers for this?

jam1garner commented:
> Are there any updates on the blockers for this?

Still waiting to hear back from Rust foundation's legal counsel

nellshamrell commented:
@skade @wesleywiser I think I remember you were having some discussions on this on the project side, do you have an update on that?

dlight commented:
> I would prefer if we could have that comment from legal before proceeding.

Is the specifics of this target substantially different from the [3ds target](https://doc.rust-lang.org/nightly/rustc/platform-support/armv6k-nintendo-3ds.html)? For example, are contributors to the `armv6k-nintendo-3ds` required to not have signed any NDA? And should the policy for this target be the same?

leo60228 commented:
The 3DS target uses an existing C-based homebrew toolchain instead of
talking directly to the system. That toolchain does have a Switch
equivalent, but I would like to avoid that for several reasons (primary
ones being a non-standard ABI requiring modifications to LLVM and the
developers being actively hostile to its usage with non-C languages or
non-GCC compilers).

On Sat, Jun 25, 2022, 4:46 AM Elias Gabriel Amaral da Silva <
***@***.***> wrote:

> I would prefer if we could have that comment from legal before proceeding.
>
> Is the specifics of this target substantially different from the 3ds
> target
> <https://doc.rust-lang.org/nightly/rustc/platform-support/armv6k-nintendo-3ds.html>?
> For example, are contributors to the armv6k-nintendo-3ds required to not
> have signed any NDA? And should the policy for this target be the same?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/88991#issuecomment-1166230446>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB7X32KMJAREUHPPTJARYJ3VQ3BM7ANCNFSM5EDLWRAA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jam1garner commented:
Additionally, the 3ds is EoL, the Switch is still actively sold. And another factor (echoing leo's sentiment but from a different angle) I think linking against a homebrew focused toolchain is likely a net negative for any use in published games (more applicable to a hypothetical nnsdk target, but also potentially applicable to a freestanding target).

I don't think there's actually any concern regarding the switch targets, just more things for project members and foundation legal to consider.

jam1garner commented:
And @wesleywiser @nellshamrell any changes?   I know @skade left the core team, is there anyone who was involved with the policy decision here that is still involved, or has this stalled and we need to find new project members to work with on this? Want to understand whether or not this is needing patience or finding new people within the rust project to discuss this with

wesleywiser commented:
@jam1garner thanks so much for your patience! I'm sorry it's taken this long to get an answer back to you. 

Per our discussions with the Foundation's legal staff, we don't see an issue with merging this PR therefore:

@bors r+ rollup

bors commented:
:pushpin: Commit 29611d2ee50418a7f7149e33bd2e5445d29dbcd1 has been approved by `wesleywiser`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=wesleywiser 29611d2ee50418a7f7149e33bd2e5445d29dbcd1 -->
<!-- homu: {"type":"Approved","sha":"29611d2ee50418a7f7149e33bd2e5445d29dbcd1","approver":"wesleywiser","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

wesleywiser commented:
Re-approving

@bors r+ rollup

bors commented:
:pushpin: Commit 62aafb01b11db42e358d81c53719727c1daca33c has been approved by `wesleywiser`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=wesleywiser 62aafb01b11db42e358d81c53719727c1daca33c -->
<!-- homu: {"type":"Approved","sha":"62aafb01b11db42e358d81c53719727c1daca33c","approver":"wesleywiser","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
